File: src/main/java/org/jdc/template/util/RxUtil.java
package org.jdc.template.util;


import android.support.annotation.Nullable;

import java.io.IOException;

import javax.annotation.Nonnull;

import io.reactivex.Single;
import retrofit2.Call;
import retrofit2.Response;
import timber.log.Timber;

public final class RxUtil {
    private RxUtil() {
    }

    // ***** Retrofit *****

    public static <T> Single<Response<T>> toRetrofitObservable(Call<T> call) {
        return Single.defer(() -> Single.just(callRetrofit(call)));
    }

    @Nullable
    public static <T> Response<T> callRetrofit(@Nonnull Call<T> call) {
        try {
            return call.execute();
        } catch (IOException e) {
            Timber.e(e, "Failed to call retrofit execute()");
        }

        return null;
    }

    @Nullable
    public static <T> T verifyRetrofitResponse(@Nullable Response<T> response) {
        if (response == null) {
            return null;
        }

        if (response.isSuccessful()) {
            return response.body();
        } else {
            return null;
        }
    }
}

File: src/main/kotlin/org/jdc/template/util/RxUtil.kt
package org.jdc.template.util


import io.reactivex.Single
import retrofit2.Call
import retrofit2.Response
import timber.log.Timber
import java.io.IOException

object RxUtil {

    // ***** Retrofit *****

    fun <T> toRetrofitObservable(call: Call<T>): Single<Response<T>> {
        return Single.defer { Single.just<Response<T>>(callRetrofit(call)) }
    }

    fun <T> callRetrofit(call: Call<T>): Response<T>? {
        try {
            return call.execute()
        } catch (e: IOException) {
            Timber.e(e, "Failed to call retrofit execute()")
        }

        return null
    }

    fun <T> verifyRetrofitResponse(response: Response<T>?): T? {
        Timber.i("Retrofit thread: ${Thread.currentThread().name}")

        if (response == null) {
            return null
        }

        if (response.isSuccessful) {
            return response.body()
        } else {
            return null
        }
    }
}
